### PR TITLE
Repairing the large-work fork warning system

### PR DESCRIPTION
**Pull request by jl2012**
The large-work fork warning system is broken under header first validation. This should fix it by storing all valid headers, even if the block contents or the previous blocks are invalid.

It also stores invalid-version-but-otherwise-valid headers. This allows a potential planned hardfork to be followed by the client (a simplified version of the hardfork bit and hardfork warning system I proposed: https://github.com/bitcoin/bips/pull/317 and https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-December/013332.html )

### CODE REVIEW

**Comment by petertodd** [comment]
Concept ACK on warning re: invalid version headers.

File: qa/rpc-tests/forkwarning.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            new_block.solve()
+            new_blocks.append(new_block)
+            tip_invalid = new_block.sha256
+            self.submit_header(new_block)
+            assert_equal(self.nodes[0].getinfo()['blocks'], 105)
+            warning = self.nodes[0].getinfo()['errors']
+            if (j < 7):
+                assert_equal(warning, BASIC_WARNING)
+            else:
+                # Warning starts when block 103 is found since it doesn't knwo the block is invalid
```
Comment (paveljanik) [inline qa/rpc-tests/forkwarning.py]: knwo -> know

File: qa/rpc-tests/forkwarning.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+                assert_equal(warning, FORK_WARNING)
+
+        # Submit the blocks. Now it knows the first block is invalid, but it doesn't show invalid chain warning because
+        # only the first block is marked BLOCK_FAILED_VALID
+        for b in new_blocks:
+            self.submit_block(b)
+            assert_equal(self.nodes[0].getinfo()['blocks'], 105)
+            assert_equal(warning, FORK_WARNING)
+
+        # Build to 185a to stop the warning
```
Comment (paveljanik) [inline qa/rpc-tests/forkwarning.py]: 185a -> 185

**Comment by laanwj** [comment]
Concept ACK, but due to the non-localized changes required to core code my opinion is that it is too risky to include in 0.14, given that it just fixes the fork check. I'd prefer to merge this after the 0.14 split-off.

**Comment by luke-jr** [comment]
To actually fix the warning system, we need to avoid banning peers that give us invalid blocks. Unfortunately, we currently also *rely* on that banning to ensure we have peers on the same network as us.

**Comment by jl2012** [comment]
@luke-jr we actually don't ban all types of invalid blocks. For example, violation of all softforks after P2SH are not banned. Also, with this PR it won't ban a node for sending header of a child of invalid block. As long as the PoW and timestamp are valid, the header chain will be followed.

File: qa/rpc-tests/forkwarning.py (hunk 2 above)
Comment (jl2012) [inline qa/rpc-tests/forkwarning.py]: yes. Maybe I could make the comments clearer

**Comment by rebroad** [comment]
@luke-jr there's a middle ground perhaps. set a maximum number of peers on different chains.

**Comment by sdaftuar** [comment]
@sipa @jl2012 and I discussed this a bit in person. I think we should revisit this PR after we have better DoS protections in place against storing arbitrarily long headers chains.  (I plan to propose one potential solution to this soon.)

**Comment by fanquake** [comment]
Closing for now. This can be revisited in future as suggested by @sdaftuar.

**Pull request closed without merging**